core/core_isolate.rs

```diff
@@ -405,47 +363,76 @@ impl CoreIsolate {
 
     snapshot
   }
+
+  /// Defines the how Deno.core.dispatch() acts.
+  /// Called whenever Deno.core.dispatch() is called in JavaScript. zero_copy_buf
+  /// corresponds to the second argument of Deno.core.dispatch().
+  ///
+  /// Requires runtime to explicitly ask for op ids before using any of the ops.
+  pub fn register_op<F>(&mut self, name: &str, op: F) -> OpId
+  where
+    F: Fn(&mut CoreIsolateState, &[u8], Option<ZeroCopyBuf>) -> Op + 'static,
+  {
+    let state_rc = Self::state(self);
+    let mut state = state_rc.borrow_mut();
+    state.op_registry.register(name, op)
+  }
 }
 
 impl Future for CoreIsolate {
   type Output = Result<(), ErrBox>;
 
   fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-    let inner = self.get_mut();
-    inner.waker.register(cx.waker());
-    inner.shared_init();
+    let core_isolate = self.get_mut();
+    core_isolate.shared_init();
 
-    let v8_isolate = inner.v8_isolate.as_mut().unwrap();
-    let js_error_create_fn = &*inner.js_error_create_fn;
-    let js_recv_cb = &inner.js_recv_cb;
-    let js_macrotask_cb = &inner.js_macrotask_cb;
-    let pending_promise_exceptions = &mut inner.pending_promise_exceptions;
+    let state_rc = Self::state(core_isolate);
+    {
+      let state = state_rc.borrow();
+      state.waker.register(cx.waker());
+    }
 
-    let mut hs = v8::HandleScope::new(v8_isolate);
+    let mut hs = v8::HandleScope::new(core_isolate);
     let scope = hs.enter();
-    let context = inner.global_context.get(scope).unwrap();
+    let context = {
+      let state = state_rc.borrow();
+      state.global_context.get(scope).unwrap()
+    };
     let mut cs = v8::ContextScope::new(scope, context);
     let scope = cs.enter();
 
-    check_promise_exceptions(
-      scope,
-      pending_promise_exceptions,
-      js_error_create_fn,
-    )?;
+    check_promise_exceptions(scope)?;
 
     let mut overflow_response: Option<(OpId, Buf)> = None;
 
     loop {
+      let mut state = state_rc.borrow_mut();
       // Now handle actual ops.
-      inner.have_unpolled_ops = false;
+      state.have_unpolled_ops = false;
+
+      let pending_r = state.pending_ops.poll_next_unpin(cx);
+      let unref_r = state.pending_unref_ops.poll_next_unpin(cx);
+      match pending_r {
```

Comment:
This is bad; if `pending_r` is always ready then it will starve `unref_r` - meaning none of `unref` ops will be returned to JS side

Reply:
Noted -  but I think we should just go with it if the tests go green. We ought to have test for that situation anyway.